cli/stats_helpers.go

```diff
@@ -0,0 +1,284 @@
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	return s.StatsEntry
+}
+
+// SetError will set error of StatsEntryWithLock
+func (s *StatsEntryWithLock) SetError(err error) {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	s.error = err
+}
+
+// GetError will return the error string of StatsEntryWithLock
+func (s *StatsEntryWithLock) GetError() string {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	return s.error.Error()
+}
+
+func collect(ctx context.Context, s *StatsEntryWithLock, cli client.CommonAPIClient, streamStats bool, waitFirst *sync.WaitGroup) {
+	logrus.Debugf("collecting stats for %s", s.container)
+	var (
+		getFirst       bool
+		previousCPU    uint64
+		previousSystem uint64
+		u              = make(chan error, 1)
+	)
+
+	defer func() {
+		// if error happens and we get nothing of stats, release wait group whatever
+		if !getFirst {
+			getFirst = true
+			waitFirst.Done()
+		}
+	}()
+
+	response, err := cli.ContainerStats(ctx, s.container, streamStats)
+	if err != nil {
+		s.SetError(err)
+		return
+	}
+	defer response.Close()
+
+	dec := json.NewDecoder(response)
+	go func() {
+		for {
+			var (
+				v                      *types.ContainerStats
+				memPercent, cpuPercent float64
+				blkRead, blkWrite      uint64
+				mem, memLimit          float64
+				pidsStatsCurrent       uint64
+			)
+
+			if err := dec.Decode(&v); err != nil {
+				dec = json.NewDecoder(io.MultiReader(dec.Buffered(), response))
+				u <- err
+				break
+			}
+
+			// first time PrecpuStats will be nil
+			if v.PrecpuStats != nil {
+				previousCPU = v.PrecpuStats.CPUUsage.TotalUsage
+				previousSystem = v.PrecpuStats.SyetemCPUUsage
+			}
+			cpuPercent = calculateCPUPercentUnix(previousCPU, previousSystem, v)
+			blkRead, blkWrite = calculateBlockIO(v.BlkioStats)
+			mem = calculateMemUsageUnixNoCache(v.MemoryStats)
+			memLimit = float64(v.MemoryStats.Limit)
+			memPercent = calculateMemPercentUnixNoCache(memLimit, mem)
+			pidsStatsCurrent = v.PidsStats.Current
+			netRx, netTx := calculateNetwork(v.Networks)
+
+			s.mutex.Lock()
+			s.name = v.Name
+			s.id = v.ID
+			s.cpuPercentage = cpuPercent
+			s.memory = mem
+			s.memoryLimit = memLimit
+			s.memoryPercentage = memPercent
+			s.networkRx = netRx
+			s.networkTx = netTx
+			s.blockRead = float64(blkRead)
+			s.blockWrite = float64(blkWrite)
+			s.pidsCurrent = pidsStatsCurrent
+			s.mutex.Unlock()
+
+			u <- nil
+			if !streamStats {
+				return
+			}
+		}
+	}()
+	for {
+		select {
+		case <-time.After(2 * time.Second):
+			// zero out the values if we have not received an update within
+			// the specified duration.
+			s.SetError(errors.New("timeout waiting for stats"))
+			// if this is the first stat you get, release WaitGroup
+			if !getFirst {
+				getFirst = true
+				waitFirst.Done()
+			}
+		case err := <-u:
+			if err != nil {
+				s.SetError(err)
+				return
+			}
+			// if this is the first stat you get, release WaitGroup
+			if !getFirst {
+				getFirst = true
```

Comment:
I think `getFirst` or other thing should be done in other goroutine. The `u` channel is used to get the error. If you want to get the notification, it is better to use other channel, not `u`. BTW, could we use `errCh` something like that. `u` contains less info so that we cannot get your point well during review.

Reply:
for the `error is nil` case, I think it can be done in first goroutine. `u`(errch) is only used to get error and exit. it can make the code clear. WDYT?

Reply:
ok, already change it.